### ISSUE TITLE
Anonymize production data for use in dev and testing environments

### ISSUE THREAD

**Issue by pdpinch** [mitodl/micromasters#2065]
_Part of #2270. This description is being updated to reflect recent discussions._

We want to use data as is it exists in production to test our app as accurately as possible, but we don't want our users' personal information to exist in multiple, possibly-less-secure databases. We should look into methods/tools for creating a postgres data dump in a way that obscures specific columns in specific tables.

The current list of fields we know we want to scrub as of 1/11/2017:

- username
- email
- first, last name
- auth tokens (we may just want to delete these entirely as part of the scrubbing process)

**Acceptance criteria:** 

1. We have a final, exhaustive list of every piece of data we want to scrub.
1. We have some proven method for dumping postgres data in a way that obscures specific columns in specific tables. If this isn't possible, we can label this as wontfix and explore other options.

Here's the original description from @pdpinch:

> Sometimes developers need production data for testing. But we don't want to proliferate any personal data. We could use a script for dumping and scrubbing data. 
> 
> data to be scrubbed:
> 
> - [ ] e-mail addresses
> - [ ] user tokens

**Comment by pdpinch** [comment]
It would also be nice if we could find a way to remove the tokens but also avoid getting lots of errors when celery tries to update the cached edx data.

**Comment by aliceriot** [comment]
Optimally this is part of the default development environment - it's nice not to have to fiddle with data setup when you're trying to test something or repro a bug.

**Comment by gsidebo** [comment]
Some reasons why it's still relevant/important to generate our own test data for manual testing:

1. Size/Speed. A prod data dump is basically guaranteed to be a great deal larger than our own test data. Reindexing on the prod dump that Tobias sent me took 73 seconds last time I ran it, and that's a single example of the possible performance penalties. I do not want that to be part of my dev workflow.
1. Precision. We cannot guarantee that prod data covers edge cases that we know and care about for testing.
1. Local development and PR review complexity. When we want to test something like dashboard course states or grades in learners search, the steps to match whatever condition is needed to develop and test a feature are typically pretty simple. That's because we understand certain things about the user and course data. Doing the same with scrubbed prod data would involve mangling user/course/etc. data to match our required conditions. That process could become complex and time-consuming.
1. Possibility of running Selenium. If we ever want to run Selenium, we absolutely cannot run it against data that we don't explicitly set up to match the conditions we care about. Tobias posted a link to a talk given by Dropbox about how they use Selenium. They use generated test data to run their Selenium tests against ([timestamped link](https://www.youtube.com/watch?v=z1aUuqKg_gA&t=22m06s)). Running against volatile prod data would be essentially impossible.
1. Security. Scrubbing the data is something that we can get wrong. I don't know if there is less of an impact if that error is limited to -rc and -ci vs. everyone's individual machines, but it's worth bringing up.

I do agree with the idea that we should test against a prod data dump on -ci and -rc, and possibly on local machines if it's explicitly needed (though in those cases I think we should set up an external DB rather than loading the data locally). If our test data is insufficient, I think we should address that insufficiency directly rather than replacing it wholesale with scrubbed prod data.

**Comment by aliceriot** [comment]
Is there any reason why both approaches can't coexist? At my last company we used scrubbed production data (well, a small subset of it) for local development and it was nice to have lots of stuff already set up and not have to worry about it (and filled in with values reflecting what users are actually doing).

I don't see why we couldn't have test data which is specifically generated to reflect all the edge cases we know of (for instance all the difference course run states) and then individual developers could switch back and forth between a copy of some of the production data and our generated test data as makes sense for the features they're working on or individual preference.

**Comment by aliceriot** [comment]
I agree that anonymizing the data is a concern, although the real risk I think is if the data were public, which it doesn't seem to me is what is being suggested. I would trust our engineering team to have access to a dump where last names, emails, DOB, and so on were randomly fuzzed.

**Comment by gsidebo** [comment]
@aliceriot i think we're advocating for roughly the same thing. keep generated data around to test edge cases (and put in the effort to keep that generated data current), and make a scrubbed prod data dump available for use in -rc/-ci and any relevant issues being worked on locally
